evidence/pool.go

```diff
@@ -190,20 +185,77 @@ func (evpool *Pool) MarkEvidenceAsCommitted(height int64, lastBlockTime time.Tim
 	// make a map of committed evidence to remove from the clist
 	blockEvidenceMap := make(map[string]struct{})
 	for _, ev := range evidence {
-		evpool.store.MarkEvidenceAsCommitted(ev)
-		blockEvidenceMap[evMapKey(ev)] = struct{}{}
+		// check that evidence has not already been committed
+		if evpool.IsCommitted(ev) {
+			continue
+		}
+		// As the evidence is stored in the block store we only need to record the height that it was saved at.
+		key := keyCommitted(ev)
+		evBytes := cdc.MustMarshalBinaryBare(height)
+		if err := evpool.evidenceStore.Set(key, evBytes); err != nil {
+			evpool.logger.Error("Unable to add committed evidence", "err", err)
+			// if we can't move evidence to committed then don't remove the evidence from pending
+			continue
+		}
+		// if pending, remove from that bucket, remember not all evidence has been seen before
+		if evpool.IsPending(ev) {
+			evpool.removePendingEvidence(ev)
+			blockEvidenceMap[evMapKey(ev)] = struct{}{}
+		}
 	}
 
 	// remove committed evidence from the clist
-	evidenceParams := evpool.State().ConsensusParams.Evidence
-	evpool.removeEvidence(height, lastBlockTime, evidenceParams, blockEvidenceMap)
+	if len(blockEvidenceMap) != 0 {
+		evidenceParams := evpool.State().ConsensusParams.Evidence
+		evpool.removeEvidenceFromList(height, lastBlockTime, evidenceParams, blockEvidenceMap)
+	}
+}
+
+func (evpool *Pool) Has(evidence types.Evidence) bool {
+	return evpool.IsPending(evidence) || evpool.IsCommitted(evidence)
 }
 
-// IsCommitted returns true if we have already seen this exact evidence and it
-// is already marked as committed.
+func (evpool *Pool) IsExpired(evidence types.Evidence) bool {
+	var (
+		params       = evpool.State().ConsensusParams.Evidence
+		ageDuration  = evpool.State().LastBlockTime.Sub(evidence.Time())
+		ageNumBlocks = evpool.State().LastBlockHeight - evidence.Height()
+	)
+	return ageNumBlocks > params.MaxAgeNumBlocks &&
+		ageDuration > params.MaxAgeDuration
+}
+
+// IsCommitted returns true if we have already seen this exact evidence and it is already marked as committed.
 func (evpool *Pool) IsCommitted(evidence types.Evidence) bool {
-	ei := evpool.store.getInfo(evidence)
-	return ei.Evidence != nil && ei.Committed
+	key := keyCommitted(evidence)
+	ok, err := evpool.evidenceStore.Has(key)
+	if err != nil {
+		evpool.logger.Error("Unable to find committed evidence", "err", err)
+	}
+	return ok
+}
+
+// Checks whether the evidence is already pending. DB errors are passed to the logger.
+func (evpool *Pool) IsPending(evidence types.Evidence) bool {
+	key := keyPending(evidence)
+	ok, err := evpool.evidenceStore.Has(key)
+	if err != nil {
+		evpool.logger.Error("Unable to find pending evidence", "err", err)
```

Comment:
I think as we are logging it and not returning the error directly back then it is okay because people are more likely to trace back the errors sometime later and it may help for debugging if they know which evidence couldn't be found.

Then again, I'm not familiar with what would cause the error because if it simply wasn't there it would return false. Perhaps it has nothing to do with the evidence itself and more the db - in that case I guess it's not necessary

Reply:
^ sorry that was a bit long - just steam of thought